parser/src/parse.rs

```diff
@@ -0,0 +1,423 @@
+                    res.push(c);
+                    continue;
+                },
+            };
+        } else if c == '\'' {
+            if in_double_quotes {
+                res.push(c);
+                continue
+            }
+
+            in_single_quotes = !in_single_quotes;
+            continue;
+        } else if c == '"' {
+            if in_single_quotes {
+                res.push(c);
+                continue
+            }
+
+            in_double_quotes = !in_double_quotes;
+            continue;
+        } else if !in_quotes {
+            return Err(ParseError::MissingOpeningQuote);
+        }
+
+        res.push(c);
+    }
+
+    Ok(res)
+}
+
+fn parse_quoted_string(s: &str, mut chars: &mut Enumerate<Chars>, mut res: String, quote: char,) -> Result<String, ParseError> {
+    let mut in_single_quotes = quote == '\'';
+    let mut in_double_quotes = quote == '"';
+
+    while let Some((idx, c)) = chars.next() {
+        let in_quotes = in_single_quotes || in_double_quotes;
+
+        if c == '\\' &&  in_quotes {
+            match chars.next() {
+                None => {
+                    return Err(ParseError::InvalidEscape {
+                        escape: format!("{}", c),
+                        index: idx,
+                        string: String::from(s),
+                    });
+                }
+                Some((idx, c2)) => {
+                    match c2 {
+                        'a' => res.push('\u{07}'),
+                        'b' => res.push('\u{08}'),
+                        'v' => res.push('\u{0B}'),
+                        'f' => res.push('\u{0C}'),
+                        'n' => res.push('\n'),
+                        'r' => res.push('\r'),
+                        't' => res.push('\t'),
+                        '\\' => res.push('\\'),
+                        '?' => res.push('?'),
+                        '\'' => {
+                            if in_double_quotes {
+                                res.push('\\');
+                                res.push('\'');
+                            } else {
+                                res.push('\'');
+                            }
+                        },
+                        '"' => {
+                            if in_single_quotes {
+                                res.push('\\');
+                                res.push('"');
+                            } else {
+                                res.push('"');
+                            }
+                        },
+                        '`' => res.push('`'),
+                        'x' => res.push(
+                            parse_unicode_hex(2, &mut chars).map_err(|x| {
+                                ParseError::InvalidUnicode {
+                                    source: x,
+                                    index: idx,
+                                    string: String::from(s),
+                                }
+                            }).unwrap()
```

Comment:
**Must Change**
These errors are being `.unwrap()`'d which means that if there's an error, it will panic the program rather than returning the error. Since this function already returns `ParseError`, let's make sure to return these errors to the caller. Adding the `?` would probably work instead of unwrap.

```rust
parse_unicode_hex(2, &mut chars).map_err(|x| {
    ParseError::InvalidUnicode {
        source: x,
        index: idx,
        string: String::from(s),
    }
})?
```

**Optional Code Style Suggestions**
As a code style suggestion, there's a lot of repetition parsing the different hex and octal codes. You might consider trying to simplify by (for example) handling all hex parsing in one match arm. This code doesn't compile as-is but you get the idea.
```rust
'x' | 'u' | 'U' => {
    let count = match c2 {
        'x' => 2,
        'u' => 4,
        'U' => 8,
    };
    res.push(parse_unicode_hex(count, chars).map_err(|source| {
        ParseError::InvalidUnicode {
            source,
            index: idx,
            string: String::from(s),
        }
    })?);
}
```

Another very small stylistic suggestion: You also might consider lifting your `res.push(...)` outside the match. If every match arm results in a push (which it would once you implement the error result handling suggested earlier), then it might be cleaner to do this
```rust
res.push(match c2 {
    'a' => '\u{07}',
    'b' => '\u{08}',
    'v' => '\u{0B}',
    'f' => '\u{0C}',
```

then this
```rust
match c2 {
    'a' => res.push('\u{07}'),
    'b' => res.push('\u{08}'),
    'v' => res.push('\u{0B}'),
    'f' => res.push('\u{0C}'),
```

Reply:
Regarding the `unwrap` - I wonder if match should return a `Result` instead? E.g.:

```rust
res.push(match c2 {
    'a' => Ok('\u{07}'),
    'b' => Ok('\u{08}'),
    'v' => Ok('\u{0B}'),
    'f' => Ok('\u{0C}'),
```

Then it's easy to handle outside `match` as well. Or do you think it's too expensive to allocate it?

Otherwise I think I have to do something like this when parsing hex/oct values:

```rust
'x' | 'u' | 'U' => {
    let length = match c2 {
        'x' => 2,
        'u' => 4,
        'U' => 8,
        _ => unreachable!(),
    };

    let result = parse_unicode_hex(length, &mut chars)
        .map_err(|x| ParseError::InvalidUnicode {
            source: x,
            index: idx,
            string: String::from(s),
        });

    match result {
        Ok(value) => value,
        Err(e) => {
            return Err(e);
        }
    }
}
```

Reply:
Can't you use the `?` operator?
```rust
parse_unicode_hex(length, &mut chars)
        .map_err(|x| ParseError::InvalidUnicode {
            source: x,
            index: idx,
            string: String::from(s),
        })?;
```

Reply:
Looks good! I'll do one final bass and then merge if you're happy with where things are at.